python/lsst/daf/butler/registry/queries/_query.py

```diff
@@ -178,6 +185,138 @@ def datasetType(self) -> Optional[DatasetType]:
+            return 0
+        sql = self.sql
+        if sql is None:
+            return 1
+        if exact and self.spatial:
+            filtered_count = 0
+            for _ in self.rows(db, region=region):
+                filtered_count += 1
+            return filtered_count
+        else:
+            return db.query(
+                sql.with_only_columns([sqlalchemy.sql.func.count()]).order_by(None)
+            ).scalar()
+
+    def any(
+        self,
+        db: Database, *,
+        region: Optional[Region] = None,
+        execute: bool = True,
+        exact: bool = True,
+    ) -> bool:
+        """Test whether this query returns any results.
+
+        Parameters
+        ----------
+        db : `Database`
+            Object managing the database connection.
+        region : `sphgeom.Region`, optional
+            A region that any result-row regions must overlap in order to be
+            yielded.  If not provided, this will be ``self.whereRegion``, if
+            that exists.
+        execute : `bool`, optional
+            If `True`, execute at least a ``LIMIT 1`` query if it cannot be
+            determined prior to execution that the query would return no rows.
+        exact : `bool`, optional
+            If `True`, run the full query and perform post-query filtering if
+            needed, until at least one result row is found.  If `False`, the
+            returned result does not account for post-query filtering, and
+            hence may be `True` even when all result rows would be filtered
+            out.
+
+        Returns
+        -------
+        any : `bool`
+            `True` if the query would (or might, depending on arguments) yield
+            result rows.  `False` if it definitely would not.
+        """
+        if self._doomed_by:
+            return False
+        sql = self.sql
+        if sql is None:
+            return True
+        if exact and not execute:
+            raise TypeError("Cannot obtain exact results without executing the query.")
+        if exact and self.spatial:
+            for _ in self.rows(db, region=region):
+                return True
+            return False
+        elif execute:
+            return db.query(sql.limit(1)).one_or_none() is not None
+        else:
+            return True
+
+    def explain_no_results(
+        self,
+        db: Database, *,
+        region: Optional[Region] = None,
+    ) -> Iterator[str]:
+        """Return human-readable messages that may help explain why the query
+        yields no results.
+
+        Parameters
+        ----------
+        db : `Database`
+            Object managing the database connection.
+        region : `sphgeom.Region`, optional
+            A region that any result-row regions must overlap in order to be
+            yielded.  If not provided, this will be ``self.whereRegion``, if
+            that exists.
+
+        Returns
+        -------
+        messages : `Iterator` [ `str` ]
```

Comment:
Should this be 

    Yields
    ------
    message : str

Reply:
I think either is correct, and it formally comes down to the subtlety of whether you want to declare that an iterator-returning function is implemented as a generator.  I generally prefer to hide that, which I gather is kind of unpythonic but better for defining abstract interfaces.